docs/prompting/thought_generation/chain_of_thought_zero_shot/tab_cot.md

```diff
@@ -1,7 +1,122 @@
 ---
-title: ""
-description: ""
-keywords: ""
+description: "Tab-CoT encourages LLMs to output reasoning as a markdown table, improving the structure and reasoning of its output"
 ---
 
-[wip]
+By getting language models to output their reasoning as a structured markdown table, we can improve their reasoning capabilities and the quality of their outputs. This is known as Tabular Chain Of Thought (Tab-CoT) <sup><a href="https://arxiv.org/pdf/2305.17812">1</a></sup>.
+
+We can implement this using `instructor` as a response object as seen below to ensure we get exactly the data that we want. Each row in our table is represented here as a `ReasoningStep` object.
+
+```python hl_lines="36-38"
+import instructor
+from openai import OpenAI
+from pydantic import BaseModel, Field
+from textwrap import dedent
+
+client = instructor.from_openai(OpenAI())
+
+
+class ReasoningStep(BaseModel):
+    step: int = Field(description="The step number")
+    subquestion: str = Field(description="Subquestion to solve")
+    procedure: str = Field(
+        description="""Any intermediate computation
+        that was done in the reasoning process. Leave
+        empty if no computation is needed""",
+    )
+    result: str
+
+
+class Response(BaseModel):
+    reasoning: list[ReasoningStep] = Field(
```

Comment:
@ivanleomk can you run some tests to check to make sure the reasoning stuff always show up before the correct_answer? we should also call out in the docs that 'correct_answer' is better than a generic variable like 'response' or 'content'

Reply:
I verified that the reasoning shows up before the correct answer by using the `Partial`.

Output
```bash
reasoning=[PartialReasoningStep(step=1, subquestion='How many loaves of bread were sold in total?', procedure='93 + 39', result='132'), PartialReasoningStep(step=2, subquestion='What is the remaining number of loaves before accounting for returns?', procedure='200 - 132', result='68'), PartialReasoningStep(step=3, subquestion='How many loaves were left after accounting for the returned loaves?', procedure='68 + 6', result='74')] correct_answer=None
reasoning=[PartialReasoningStep(step=1, subquestion='How many loaves of bread were sold in total?', procedure='93 + 39', result='132'), PartialReasoningStep(step=2, subquestion='What is the remaining number of loaves before accounting for returns?', procedure='200 - 132', result='68'), PartialReasoningStep(step=3, subquestion='How many loaves were left after accounting for the returned loaves?', procedure='68 + 6', result='74')] correct_answer=None
reasoning=[PartialReasoningStep(step=1, subquestion='How many loaves of bread were sold in total?', procedure='93 + 39', result='132'), PartialReasoningStep(step=2, subquestion='What is the remaining number of loaves before accounting for returns?', procedure='200 - 132', result='68'), PartialReasoningStep(step=3, subquestion='How many loaves were left after accounting for the returned loaves?', procedure='68 + 6', result='74')] correct_answer=None
reasoning=[PartialReasoningStep(step=1, subquestion='How many loaves of bread were sold in total?', procedure='93 + 39', result='132'), PartialReasoningStep(step=2, subquestion='What is the remaining number of loaves before accounting for returns?', procedure='200 - 132', result='68'), PartialReasoningStep(step=3, subquestion='How many loaves were left after accounting for the returned loaves?', procedure='68 + 6', result='74')] correct_answer=74
reasoning=[PartialReasoningStep(step=1, subquestion='How many loaves of bread were sold in total?', procedure='93 + 39', result='132'), PartialReasoningStep(step=2, subquestion='What is the remaining number of loaves before accounting for returns?', procedure='200 - 132', result='68'), PartialReasoningStep(step=3, subquestion='How many loaves were left after accounting for the returned loaves?', procedure='68 + 6', result='74')] correct_answer=74
```


Code
```python
import instructor
from openai import OpenAI
from pydantic import BaseModel, Field
from textwrap import dedent


client = instructor.from_openai(OpenAI())


class ReasoningStep(BaseModel):
    step: int = Field(description="The step number")
    subquestion: str = Field(description="Subquestion to solve")
    procedure: str = Field(
        description="""Any intermediate computation
        that was done in the reasoning process. Leave
        empty if no computation is needed""",
    )
    result: str


class Response(BaseModel):
    reasoning: list[ReasoningStep] = Field(
        description="reasoning steps to derive answer",
    )
    correct_answer: int


def generate_structured_reasoning_response(query: str, context: str):
    response = client.chat.completions.create(
        model="gpt-4o",
        response_model=instructor.Partial[Response],
        messages=[
            {
                "role": "system",
                "content": dedent(
                    f"""
                You are an expert Question Answering system. Make sure
                to output your reasoning in structured reasoning steps
                before generating a response to the user's query.


                Context:
                {context}

                Query:
                {query}
                """
                ),
            },
        ],
        stream=True,
    )
    return response


if __name__ == "__main__":
    query = "How many loaves of bread did they have left?"
    context = """
    The bakers at the Beverly Hills Bakery baked
    200 loaves of bread on Monday morning. They
    sold 93 loaves in the morning and 39 loaves
    in the afternoon. A grocery store returned 6
    unsold loaves.
    """

    for response in generate_structured_reasoning_response(query, context):
        print(response)
```